### PR TITLE
Add support for lazy async generators

### LINKED ISSUE #5: Support async generators
The current behavior is to raise a SyntaxError if the decorator is called on an async generator function.

To support them well, the cached result must be API-compatible with that of the wrapped async generator.

### PR DESCRIPTION
**Pull request by mattalbr**
Fixes #5

### CODE REVIEW

File: once_test.py [hunk 1]
```diff
@@ -368,19 +368,98 @@ async def counting_fn2():
...
-                    yield i
 
-        # self.assertEqual([i async for i in async_yielding_iterator()], [1, 0, 1, 2])
-        # self.assertEqual([i async for i in async_yielding_iterator()], [1, 0, 1, 2])
+        gen_1 = async_yielding_iterator()
+        gen_2 = async_yielding_iterator()
+        gen_3 = async_yielding_iterator()
+
+        self.assertEqual(counter.value, 0)
+        self.assertEqual(await anext(gen_1), 1)
```
Comment (aebrahim) [inline once_test.py:431]: It looks like `anext` was introduced in python 3.10.

Maybe we add a manaul backport?

```
if sys.version_info.minor < 10:
  async def anext(iter): ...
```

**Comment by aebrahim** [comment]
It looks like mypy and black are unhappy

File: once_test.py (hunk 1 above)
Comment (mattalbr) [reply once_test.py:445]: Done!

**Comment by mattalbr** [comment]
FYI I updated the implementation to not block other generators while awaiting the next value, please give it a close look as it's definitely a bit more complicated!

**Review by aebrahim — APPROVED** [review]

**Pull request merged**